### ISSUE TITLE
jax.numpy.ldexp output does not match np.ldexp

### ISSUE THREAD

**Issue by jakevdp** [google/jax#4053]
For example:
```python
>>> import numpy as np; np.ldexp(1, 1)                                                                        
2.0
>>> import jax.numpy as jnp; jnp.ldexp(1, 1)                                                                  
DeviceArray(1., dtype=float32)
```

**Comment by PhilippThoelke** [comment]
I would like to work on this issue. The problem appears to be that [`_result_dtype(np.ldexp, x1, x2)`](https://github.com/google/jax/blob/5fcca18424e08b5d8cabac0496a8b77d6a19be6e/jax/numpy/lax_numpy.py#L746) returns `float64` and since jax does not support double precision by default, the cast [`lax.bitcast_convert_type(x, int_type)`](https://github.com/google/jax/blob/5fcca18424e08b5d8cabac0496a8b77d6a19be6e/jax/numpy/lax_numpy.py#L758) fails to convert `x` (float32) to an int64 (`int_type` is `np.int64` because `_result_dtype(np.ldexp, x1, x2)` returns `float64`).

This line [`x2 += ((x >> info.nmant) & mask) - bias`](https://github.com/google/jax/blob/5fcca18424e08b5d8cabac0496a8b77d6a19be6e/jax/numpy/lax_numpy.py#L759) is supposed to shift `x` right by the number of bits of the mantissa of a float with the same number of bits as `x` but the `info` object comes from `finfo(float64)` and `x` is an int32.

Setting the environment variable `JAX_ENABLE_X64=True` or setting the `jax_enable_x64` config flag to true fixes the problem and results match those from NumPy.

What is the preferred way of handling such situations?

Overwriting the result of `_result_dtype(np.ldexp, x1, x2)` in case the `jax_enable_x64` config flag is false would be one working solution:
```python
if config.read('jax_enable_x64'):
  dtype = _result_dtype(np.ldexp, x1, x2)
else:
  dtype = np.float32
```
instead of
```python
dtype = _result_dtype(np.ldexp, x1, x2)
```

**Comment by jakevdp** [comment]
The best thing to do would probably be to use `dtype = dtypes.canonicalize_dtype(_result_dtype("ldexp", x1, x2)`. Then that should represent the correct final dtype.
